### PR TITLE
Disable tracking protection while blok is enabled

### LINKED ISSUE #1333: Toggle Native Tracking Protection Pref off if user installs Test Pilot Tracking Protection
Related to #944, and we may not need a separate issue here, but we can cover this in triage today.

### PR DESCRIPTION
**Pull request by dannycoates**
Fixes #1333

This doesn't cover the edge cases where a user disables Test Pilot from about:addons then disables or removes Blok.

### CODE REVIEW

File: addon/index.js [hunk 1]
```diff
@@ -180,11 +181,6 @@ function installExperiment(experiment) {
 }
 
 function uninstallSelf() {
-  // First, kick out all the experiment add-ons
-  Object.keys(store.installedAddons).forEach(id => {
-    uninstallExperiment({addon_id: id});
-  });
```
Comment (dannycoates) [inline addon/index.js:186]: This is already handled in `onUnload`

File: addon/index.js [hunk 2]
```diff
@@ -361,6 +362,7 @@ exports.onUnload = function(reason) {
     if (store.installedAddons) {
       Object.keys(store.installedAddons).forEach(id => {
         uninstallExperiment({addon_id: id});
+        ExperimentHacks.uninstalled(id)
```
Comment (dannycoates) [inline addon/index.js]: The `addonListener` won't get triggered at this point, so we need to do it manually

File: addon/lib/experiment-hacks.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+const PrefsService = require('sdk/preferences/service');
+const store = require('sdk/simple-storage').storage;
+
+const BLOK_ID = 'blok@mozilla.org';
+const TP_PREFS = [
+  'privacy.trackingprotection.enabled',
+  'privacy.trackingprotection.pbmode.enabled',
+  'services.sync.prefs.sync.privacy.trackingprotection.enabled',
+  'services.sync.prefs.sync.privacy.trackingprotection.pbmode.enabled'
```
Comment (dannycoates) [inline addon/lib/experiment-hacks.js:15]: Sync prefs weren't in the original bug, but we probably want to turn off syncing of these prefs too. @clouserw @johngruen ?

**Review by lmorchard — APPROVED** [review]
:+1: with some nits / questions:
- Could use some tests for the new ExperimentHacks module
- We're adding hacks to the main add-on to disable tracking protection when blok is installed... what happens with users whose Test Pilot add-ons haven't upgraded yet?

**Comment by lmorchard** [comment]
Just thinking out loud if this ends up being a problem: Might make sense to add a bit to the hack that checks after Test Pilot upgrade or browser restart if Blok is installed and turns off tracking protection. Probably a tiny edge case not worth worrying about though

**Comment by chuckharmston** [comment]
Meta-question: why build this as a one-time-use hack, when we could store this information in the yaml for the experiment and handle it more globally?

``` yaml
prefs:
- name: something.whatever
  type: boolean
  enable: true
  disable: false
- name: something.whateverelse
  type: string
  enable: hello
  disable: goodbye
```

This is surely not a unique use case, and this is exactly the sort of thing Test Pilot should be federating for experiments, right?

**Comment by meandavejustice** [comment]
👍 

@chuckharmston That's a good point, personally I would wait until there is another addon that needs to set prefs before including it in the yaml.

The upgrade case that @lmorchard mentions is valid. That seems like a simple enough check in `exports.unLoad` in `index.js`

**Comment by dannycoates** [comment]
> Might make sense to add a bit to the hack that checks after Test Pilot upgrade or browser restart if Blok is installed and turns off tracking protection

Ok, I moved the logic for when hacks get triggered that will catch this case. Expect tests shortly

**Comment by dannycoates** [comment]
tests added

File: addon/lib/experiment-hacks.js (hunk 3 above)
Comment (ghost) [reply addon/lib/experiment-hacks.js:15]: Seems right.  If they are set I assume Firefox would keep turning it back on

**Comment by lmorchard** [comment]
Taking another look...

File: addon/test/test-experiment-hacks.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
+/*
```
Comment (lmorchard) [inline addon/test/test-experiment-hacks.js:1]: :+1: Thanks for circling back to do tests!

**Review by lmorchard — APPROVED** [review]
Looks good to me and tests make me feel warm & fuzzy

**Comment by lmorchard** [comment]
> This is surely not a unique use case, and this is exactly the sort of thing Test Pilot should be federating for experiments, right?

Yeah, for this I would say we optimize that when we have a 2nd one that needs the same thing. I also suspect we may need to keep this "hacks" module around in case future experiments need different weird things done for them that can only be done safely/securely/sanely from the main Test Pilot add-on.
